post-merge-review.sh

```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+
+set -e
+
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+NC='\033[0m'
+REGEX_REF='^\(cherry picked from commit ([a-f0-9]{40})\)$'
+
+do_reference_check() {
+  if ! [[ $(git log --format=%B "$line^!" | grep -E "$REGEX_REF") =~ $REGEX_REF ]]; then
+    echo -e "${RED}FAIL${NC}" >> "$GITHUB_OUTPUT"
+    dirty=1
+    echo "  Cannot find reference in commit message" >> "$GITHUB_OUTPUT"
+    echo "  Hint: use -x with git cherry-pick" >> "$GITHUB_OUTPUT"
+    return;
+  fi
+
+  if ! eval "$(git branch --contains "${BASH_REMATCH[1]}" "$check_branch" 2>/dev/null)"; then
+    echo -e "${RED}FAIL${NC}" >> "$GITHUB_OUTPUT"
+    dirty=1
+    echo "  Cannot find ${BASH_REMATCH[1]} from $line in $check_branch" >> "$GITHUB_OUTPUT"
+    return;
+  fi
+  echo -e "${GREEN}OK${NC}" >> "$GITHUB_OUTPUT"
+}
+
+do_subject_check() {
+  summary=$(git log --format=%s "$line^!")
+  if (($(git log --format=%s "$check_branch" | grep -Fxc "$summary") == 0)); then
+    echo -e "${RED}FAIL${NC}" >> "$GITHUB_OUTPUT"
+    dirty=1
+    echo "  Cannot find $summary in $check_branch" >> "$GITHUB_OUTPUT"
+    return;
+  fi
+  echo -e "${GREEN}OK${NC}" >> "$GITHUB_OUTPUT"
+}
+
+main() {
+  if [[ -z "${GITHUB_HEAD_REF}" ]]; then
+    echo "Must be run on pull_request or pull_request_target triggers only!" >> "$GITHUB_OUTPUT"
+    exit 1
+  fi
+
+  if [ "$GITHUB_BASE_REF" = "$2" ]; then
+    echo "Check cannot run against defined main branch!" >> "$GITHUB_OUTPUT"
+    exit 1
+  fi
+
+  commit_range="$(git merge-base "$GITHUB_BASE_REF" "$GITHUB_HEAD_REF")..$GITHUB_HEAD_REF"
+  echo "Checking range $commit_range which contains $(git log --oneline "$commit_range" | wc -l) commits" >> "$GITHUB_OUTPUT"
+  commits=$(git log --format=format:%H "$commit_range")
+
+  dirty=0
+  check_branch=$2
+
+  for line in $commits; do
+    echo -n "$(git log --format=format:"Checking %h %s..." "$line^!")" >> "$GITHUB_OUTPUT"
+    if [ "$1" = "true" ]; then
+      do_reference_check
+    else
+      do_subject_check
+    fi
+  done
+
+  exit $dirty
+}
+
+main "$@"
```

Comment:
```suggestion
main "$@" >>"$GITHUB_OUTPUT"
```
and now you have a single line to change for testing locally

Reply:
I just overwrite the var to stderr, how do I print from my sub functions if the are returning values?

Reply:
I'm not sure I understand the problem.

It's true that you cannot have one function performing both logging _and_ returning some string value on stdout at the same time, functions can only use stdout for a single purpose.

But right now you have only two functions and they only do logging, none of them returns any string (they're not even returning an exit code either but I digress)

Reply:
OK now the functions are doing `echo 0` and `echo 1` in the newer https://github.com/cujomalainey/cherry-pick-check/blob/e8e1e4fd577d/entrypoint.sh just to prove me wrong :-)

You should never do `echo 0` and `echo 1` when you can just do `return 0` and `return 1`. Then you can keep stdout for logging - and redirect the entire `main` function.

```shell

do_reference_check()
{
   ...
   return 1
}

main()
{
   ...
    ret=0
    if [ "$req_ref" = "true" ]; then
       do_reference_check || ret=1
    else
       do_subject_check || ret=1
    fi
```
or:
```shell
    if [ "$req_ref" = "true" ]; then
       do_reference_check || DIRTY_PR=$((DIRTY_PR+$?))
    else
       do_subject_check || DIRTY_PR=$((DIRTY_PR+$?))
    fi
```
or, even fancier

```shell
    if [ "$req_ref" = "true" ]; then
       do_check=do_reference_check
    else
       do_check=do_subject_check
    fi

    $do_check || DIRTY_PR=$((DIRTY_PR+$?))
```

Up to you.

Reply:
>  `if [ "$req_ref" = "true" ]; then`

If you can make absolutely sure that `$req_ref` is either `true` or `false` (e.g.: because it's an internal variable and not an input), then you can invoke them as commands directly with the much simpler:

`if $req_ref; then`